index.js

```diff
@@ -25,27 +25,53 @@ function compose (middleware) {
   }
 
   /**
+   * Make an iterator for middleware.
+   */
+
+  middleware[Symbol.iterator] = function () {
+    const self = this
+    const length = this.length
+    let i = 0
+    let context
+    let nextFunc
+
+    return {
+      next (c, n) {
+        if (!context) context = c
+        if (!nextFunc) nextFunc = n
+
+        let fn = self[i++] || nextFunc
+        let done = i > length
+        let value
+
+        if (done) {
+          value = fn ? fn(context) : void 0
+        } else {
+          value = fn(context, this.next.bind(this))
+        }
+
+        return {
+          value,
+          done
+        }
+      }
+    }
+  }
+
+  // Alternative iteration.
+  const iter = middleware[Symbol.iterator]()
+
+  /**
    * @param {Object} context
    * @return {Promise}
    * @api public
    */
 
   return function (context, next) {
-    // last called middleware #
-    let index = -1
-    return dispatch(0)
-    function dispatch (i) {
-      if (i <= index) return Promise.reject(new Error('next() called multiple times'))
-      index = i
-      const fn = middleware[i] || next
-      if (!fn) return Promise.resolve()
-      try {
-        return Promise.resolve(fn(context, function next () {
-          return dispatch(i + 1)
-        }))
-      } catch (err) {
-        return Promise.reject(err)
-      }
+    try {
+      return Promise.resolve(iter.next(context, next).value)
+    } catch (err) {
+      return Promise.reject(err)
     }
```

Comment:
How about this? `return Promise.then(() => iter.next(context, next).value)`

That removes the need for a try-catch.